instrumentation/c3p0-0.9/library/src/main/java/io/opentelemetry/instrumentation/c3p0/ConnectionPoolMetrics.java

```diff
@@ -0,0 +1,122 @@
+      dataSourceMetrics = new ConcurrentHashMap<>();
+
+  public static void registerMetrics(OpenTelemetry openTelemetry, PooledDataSource dataSource) {
+    dataSourceMetrics.compute(
+        new IdentityDataSourceKey(dataSource),
+        (key, existingCounters) ->
+            ConnectionPoolMetrics.createMeters(openTelemetry, key, existingCounters));
+  }
+
+  private static List<ObservableLongUpDownCounter> createMeters(
+      OpenTelemetry openTelemetry,
+      IdentityDataSourceKey key,
+      List<ObservableLongUpDownCounter> existingCounters) {
+    // remove old counters from the registry in case they were already there
+    removeMetersFromRegistry(existingCounters);
+
+    PooledDataSource dataSource = key.dataSource;
+
+    DbConnectionPoolMetrics metrics =
+        DbConnectionPoolMetrics.create(
+            openTelemetry, INSTRUMENTATION_NAME, dataSource.getDataSourceName());
+
+    return Arrays.asList(
+        metrics.usedConnections(wrapThrowingSupplier(dataSource::getNumBusyConnectionsDefaultUser)),
+        metrics.idleConnections(wrapThrowingSupplier(dataSource::getNumIdleConnectionsDefaultUser)),
+        metrics.pendingRequestsForConnection(
+            wrapThrowingSupplier(dataSource::getNumThreadsAwaitingCheckoutDefaultUser)));
+  }
+
+  public static void unregisterMetrics(PooledDataSource dataSource) {
+    List<ObservableLongUpDownCounter> meters =
+        dataSourceMetrics.remove(new IdentityDataSourceKey(dataSource));
+    removeMetersFromRegistry(meters);
+  }
+
+  private static void removeMetersFromRegistry(
+      @Nullable List<ObservableLongUpDownCounter> observableInstruments) {
+    if (observableInstruments != null) {
+      for (ObservableLongUpDownCounter observable : observableInstruments) {
+        observable.close();
+      }
+    }
+  }
+
+  /**
+   * A wrapper over {@link PooledDataSource} that implements identity comparison in its {@link
+   * #equals(Object)} and {@link #hashCode()} methods.
+   */
+  static final class IdentityDataSourceKey {
+    final PooledDataSource dataSource;
+
+    IdentityDataSourceKey(PooledDataSource dataSource) {
+      this.dataSource = dataSource;
+    }
+
+    @Override
+    @SuppressWarnings("ReferenceEquality")
+    public boolean equals(@Nullable Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      IdentityDataSourceKey that = (IdentityDataSourceKey) o;
+      return dataSource == that.dataSource;
+    }
+
+    @Override
+    public int hashCode() {
+      return System.identityHashCode(dataSource);
+    }
+
+    @Override
+    public String toString() {
+      return dataSource.toString();
+    }
+  }
+
+  static LongSupplier wrapThrowingSupplier(DataSourceIntSupplier supplier) {
+    return () -> {
+      try {
+        return supplier.getAsInt();
+      } catch (SQLException e) {
+        return 0;
```

Comment:
@jack-berg any thoughts on recording `0` vs not recording any value in this case?

Reply:
I'm thinking it may be better to re-throw an unchecked exception here instead of recording `0`.

It looks like it will get caught and logged here: https://github.com/open-telemetry/opentelemetry-java/blob/f280f278be0222f24b7ad7a8dc7ffc293358785b/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/CallbackRegistration.java#L100-L103

Or if we think this is common and don't want to log, we could consider suppressing it here: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/fae88de680ccbfbac79550f3b151fafc493e04b3/instrumentation-api-semconv/src/main/java/io/opentelemetry/instrumentation/api/metrics/db/DbConnectionPoolMetrics.java#L65

Reply:
Probably don't want instrumentation to be in the habit of throwing exceptions, even if the SDK can handle it. Better to have the handling to be explicit in the callback.

I do think that recording nothing is more accurate than 0.

Reply:
The problem is that the only ways to record nothing with the current implementation of `DbConnectionPoolMetrics` would be to either throw an exception, or change `DbConnectionPoolMetrics` to use providers that return a `Long` instead that can be `null`. I changed it throw since I didn't see the last comment yet, but if you can suggest a better way, I can change it.

Reply:
Do you expect exception to be thrown here on a regular basis? If so, it is likely to make user logs noisy. For reference, [here's](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/CallbackRegistration.java#L100) the code that handles these exceptions.

Reply:
I looked briefly at the c3p0 implementation, and I couldn't really find any case where these exceptions would really be thrown, so I think it's probably ok